File: RudderSDKCore/RudderSDKCore/DBPersistentManager.m
//
//  DBPersistentManager.m
//  RudderSDKCore
//
//  Created by Arnab Pal on 17/10/19.
//  Copyright © 2019 Rudderlabs. All rights reserved.
//

#import "DBPersistentManager.h"
#import "RudderLogger.h"

@implementation DBPersistentManager

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self createDB];
        [self createSchema];
    }
    return self;
}

- (void)createDB {
    if (sqlite3_open_v2([Utils getDBPath], &(self->_database), SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE | SQLITE_OPEN_FULLMUTEX, nil) == SQLITE_OK) {
        // opened correctly
    }
}

- (void)createSchema {
    NSString *createTableSQLString = @"CREATE TABLE IF NOT EXISTS events( id INTEGER PRIMARY KEY AUTOINCREMENT, message TEXT NOT NULL, updated INTEGER NOT NULL);";
    [RudderLogger logDebug:[[NSString alloc] initWithFormat:@"CreateTableSchema: %@", createTableSQLString]];
    const char* createTableSQL = [createTableSQLString UTF8String];
    sqlite3_stmt *createTableStmt = nil;
    if (sqlite3_prepare_v2(self->_database, createTableSQL, -1, &createTableStmt, nil) == SQLITE_OK) {
        if (sqlite3_step(createTableStmt) == SQLITE_DONE) {
            // table created
            [RudderLogger logDebug:@"DB Schema created"];
        } else {
            // table creation error
            [RudderLogger logError:@"DB Schema creation error"];
        }
    } else {
        // wrong statement
    }
    sqlite3_finalize(createTableStmt);
}

- (void)saveEvent:(NSString *)message {
    NSString *insertSQLString = [[NSString alloc] initWithFormat:@"INSERT INTO events (message, updated) VALUES ('%@', %ld);", message, [Utils getTimeStampLong]];
    [RudderLogger logDebug:[[NSString alloc] initWithFormat:@"saveEventSQL: %@", insertSQLString]];
    const char* insertSQL = [insertSQLString UTF8String];
    sqlite3_stmt *insertStmt = nil;
    if (sqlite3_prepare_v2(self->_database, insertSQL, -1, &insertStmt, nil) == SQLITE_OK) {
        if (sqlite3_step(insertStmt) == SQLITE_DONE) {
            // table created
            [RudderLogger logDebug:@"Event inserted to table"];
        } else {
            // table creation error
            [RudderLogger logError:@"Event insertion error"];
        }
    } else {
        // wrong statement
    }
    sqlite3_finalize(insertStmt);
}

- (void)clearEventFromDB:(int)messageId {
    
}

- (void)clearEventsFromDB:(NSArray<NSString *> *)messageIds {
    NSMutableString *messageIdsCsv = [[NSMutableString alloc] init];
    for (int index = 0; index < messageIds.count; index++) {
        [messageIdsCsv appendString:messageIds[index]];
        if (index != messageIds.count -1) {
            [messageIdsCsv appendString:@","];
        }
    }
    
    NSString *deleteSqlString = [[NSString alloc] initWithFormat:@"DELETE FROM events WHERE id IN (%@);", messageIdsCsv];
    [RudderLogger logDebug:[[NSString alloc] initWithFormat:@"deleteEventSql: %@", deleteSqlString]];
    const char* deleteSql = [deleteSqlString UTF8String];
    sqlite3_stmt *deleteStmt = nil;
    if (sqlite3_prepare_v2(self->_database, deleteSql, -1, &deleteStmt, nil) == SQLITE_OK) {
        if (sqlite3_step(deleteStmt) == SQLITE_DONE) {
            // delete successful
            [RudderLogger logDebug:@"Events deleted from DB"];
        } else {
            // delete failed
            [RudderLogger logError:@"Event deletion error"];
        }
    } else {
        // wrong statement
    }
    
    sqlite3_finalize(deleteStmt);
}

- (RudderDBMessage *)fetchEventsFromDB:(int)count {
    NSString *querySQLStirng = [[NSString alloc] initWithFormat:@"SELECT * FROM events ORDER BY updated ASC LIMIT %d;", count];
    [RudderLogger logDebug:[[NSString alloc] initWithFormat:@"fetchEventSql: %@", querySQLStirng]];
    const char* querySQL = [querySQLStirng UTF8String];
    NSMutableArray<NSString *> *messageIds = [[NSMutableArray alloc] init];
    NSMutableArray<NSString *> *messages = [[NSMutableArray alloc] init];
    
    sqlite3_stmt *queryStmt = nil;
    if (sqlite3_prepare_v2(self->_database, querySQL, -1, &queryStmt, nil) == SQLITE_OK) {
        [RudderLogger logDebug:@"events fetched from DB"];
        while (sqlite3_step(queryStmt) == SQLITE_ROW) {
            int messageId = sqlite3_column_int(queryStmt, 0);
            const unsigned char* queryResultCol1 = sqlite3_column_text(queryStmt, 1);
            NSString *message = [[NSString alloc] initWithUTF8String:(char *)queryResultCol1];
            [messageIds addObject:[[NSString alloc] initWithFormat:@"%d", messageId]];
            [messages addObject:message];
        }
    } else {
        // wrong statement
        [RudderLogger logError:@"event fetching error"];
    }
    
    RudderDBMessage *dbMessage = [[RudderDBMessage alloc] init];
    dbMessage.messageIds = messageIds;
    dbMessage.messages = messages;
    
    return dbMessage;
}

- (int)getDBRecordCount {
    NSString *countSQLString = @"SELECT COUNT(*) FROM 'events'";
    
    [RudderLogger logDebug:[[NSString alloc] initWithFormat:@"countSQLStirng: %@", countSQLString]];
    
    int count = 0;
    const char* countSQL = [countSQLString UTF8String];
    
    sqlite3_stmt *countStmt = nil;
    if (sqlite3_prepare_v2(self->_database, countSQL, -1, &countStmt, nil) == SQLITE_OK) {
        [RudderLogger logDebug:@"count fetched from DB"];
        while (sqlite3_step(countStmt) == SQLITE_ROW) {
            count = sqlite3_column_int(countStmt, 0);
        }
    } else {
        // wrong statement
        [RudderLogger logError:@"count fetching error"];
    }
    
    return count;
}

- (void)deleteAllEvents {
    
}


@end

File: RudderSampleProject/RudderSdkCore/DBPersistentManager.swift
//
//  DBPersistentManager.swift
//  RudderPlugin_iOS
//
//  Created by Arnab Pal on 14/09/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//

import Foundation
import SQLite3

class DBPersistentManager {
    private static var instance: DBPersistentManager? = nil
    private var dbInstance: OpaquePointer? = nil
    
    static func getInstance() -> DBPersistentManager {
        if (instance == nil) {
            instance = DBPersistentManager()
        }
        return instance!
    }
    
    private init() {
        createDB()
        createSchema()
    }
    
    private func createDB() {
        if (dbInstance == nil) {
            sqlite3_shutdown()
            
            if (sqlite3_open_v2(Utils.getPath(fileName: "rl_persistance.db"), &self.dbInstance, SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE | SQLITE_OPEN_FULLMUTEX, nil) == SQLITE_OK) {
                RudderLogger.logDebug(message: "DB created")
            } else {
                RudderLogger.logError(message: "DB could not be created")
            }
        }
    }
 
    /*
     * create table initially if not exists
     * */
    private func createSchema() {
        let createTableString = """
            CREATE TABLE IF NOT EXISTS events(
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            message TEXT NOT NULL,
            updated INTEGER NOT NULL);
        """
        var createTableStatement: OpaquePointer? = nil
        if sqlite3_prepare_v2(self.dbInstance, createTableString, -1, &createTableStatement, nil) == SQLITE_OK {
            if sqlite3_step(createTableStatement) == SQLITE_DONE {
                 RudderLogger.logDebug(message: "Schema created")
            } else {
                RudderLogger.logError(message: "Schema could not be created")
            }
        } else {
            RudderLogger.logError(message: "CREATE TABLE statement could not be prepared")
        }
        sqlite3_finalize(createTableStatement)
    }
    
    /*
     * save individual messages to DB
     * */
    func saveEvent(messageJson: String) {
        let insertStatementString = "INSERT INTO events (message, updated) VALUES ('"+messageJson+"', "+String(Utils.gettimeStampLong())+");"
        var insertStatement: OpaquePointer? = nil
        if sqlite3_prepare_v2(self.dbInstance, insertStatementString, -1, &insertStatement, nil) == SQLITE_OK {
            if sqlite3_step(insertStatement) == SQLITE_DONE {
                RudderLogger.logDebug(message: "Successfully inserted event")
            } else {
                RudderLogger.logError(message: "Could not insert row")
            }
        } else {
            RudderLogger.logError(message: "INSERT statement could not be prepared")
        }
        sqlite3_finalize(insertStatement)
    }
    
    /*
     * remove selected events from persistence database storage
     * */
    func clearEventsFromDB(messageIds: [Int32]) {
        var messageIdCsv = ""
        for index in 0..<messageIds.count {
            messageIdCsv.append(String(messageIds[index]))
            if (index != messageIds.count-1) {
                messageIdCsv.append(contentsOf: ",")
            }
        }
        let deleteStatementStirng = "DELETE FROM events WHERE id IN ("+messageIdCsv+");"
        var deleteStatement: OpaquePointer? = nil
        if sqlite3_prepare_v2(self.dbInstance, deleteStatementStirng, -1, &deleteStatement, nil) == SQLITE_OK {
            if sqlite3_step(deleteStatement) == SQLITE_DONE {
                RudderLogger.logDebug(message: "Successfully deleted messages")
            } else {
                RudderLogger.logError(message: "Could not delete messages")
            }
        } else {
            RudderLogger.logError(message: "DELETE statement could not be prepared")
        }
        sqlite3_finalize(deleteStatement)
    }
    
    
    func fetchEventsFromDB(count: Int32) -> RudderDBMessage {
        let queryStatementString = "SELECT * FROM events ORDER BY updated ASC LIMIT "+String(count)
        var messageIds: [Int32] = []
        var messages: [String] = []
        var queryStatement: OpaquePointer? = nil
        if sqlite3_prepare_v2(self.dbInstance, queryStatementString, -1, &queryStatement, nil) == SQLITE_OK {
            while (sqlite3_step(queryStatement) == SQLITE_ROW) {
                let messageId = sqlite3_column_int(queryStatement, 0)
                let queryResultCol1 = sqlite3_column_text(queryStatement, 1)
                let message = String(cString: queryResultCol1!)
                messageIds.append(messageId)
                messages.append(message)
            }
        } else {
            RudderLogger.logError(message: "SELECT statement could not be prepared")
        }
        sqlite3_finalize(queryStatement)
        
        return RudderDBMessage(messageIds: messageIds, messages: messages)
    }
    
    func getDBRecordCount() -> Int32 {
        var count: Int32 = 0
        let countQuereStatementString = "SELECT COUNT(*) FROM 'events';"
        var queryStatement: OpaquePointer? = nil
        if sqlite3_prepare_v2(self.dbInstance, countQuereStatementString, -1, &queryStatement, nil) == SQLITE_OK {
            while (sqlite3_step(queryStatement) == SQLITE_ROW) {
                count = sqlite3_column_int(queryStatement, 0)
            }
        } else {
            RudderLogger.logError(message: "SELECT statement could not be prepared")
        }
        sqlite3_finalize(queryStatement)
        
        return count
    }
}
